cmd/influx_inspect/dumptsmwal/dumptsmwal.go

```diff
@@ -0,0 +1,161 @@
+// Package dumptsmwal dumps all data from a WAL file.
+package dumptsmwal
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"sort"
+
+	"github.com/influxdata/influxdb/tsdb/engine/tsm1"
+)
+
+// Command represents the program execution for "influxd dumptsmwal".
+type Command struct {
+	// Standard input/output, overridden for testing.
+	Stderr io.Writer
+	Stdout io.Writer
+
+	showDuplicates bool
+}
+
+// NewCommand returns a new instance of Command.
+func NewCommand() *Command {
+	return &Command{
+		Stderr: os.Stderr,
+		Stdout: os.Stdout,
+	}
+}
+
+// Run executes the command.
+func (cmd *Command) Run(args ...string) (err error) {
+	fs := flag.NewFlagSet("dumptsmwal", flag.ExitOnError)
+	fs.SetOutput(cmd.Stdout)
+	fs.BoolVar(&cmd.showDuplicates, "show-duplicates", false, "prints keys with out-of-order or duplicate values")
+	fs.Usage = cmd.printUsage
+	if err := fs.Parse(args); err != nil {
+		return err
+	} else if fs.NArg() == 0 {
+		fmt.Printf("path required\n\n")
+		fs.Usage()
+		return nil
+	}
+
+	// Process each TSM WAL file.
+	for _, path := range fs.Args() {
+		if err := cmd.process(path); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (cmd *Command) process(path string) error {
+	if filepath.Ext(path) != "."+tsm1.WALFileExtension {
+		log.Printf("invalid wal filename, skipping %s", path)
+	}
+
+	// Track the earliest timestamp for each key and a set of keys with out-of-order points.
+	minTimestampByKey := make(map[string]int64)
+	duplicateKeys := make(map[string]struct{})
+
+	// Open WAL reader.
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	r := tsm1.NewWALSegmentReader(f)
+
+	// Iterate over the WAL entries.
+	for r.Next() {
+		entry, err := r.Read()
+		if err != nil {
+			return fmt.Errorf("cannot read entry: %s", err)
+		}
+
+		switch entry := entry.(type) {
+		case *tsm1.WriteWALEntry:
+			if !cmd.showDuplicates {
+				fmt.Printf("[write] sz=%d\n", entry.MarshalSize())
+			}
+
+			keys := make([]string, 0, len(entry.Values))
+			for k := range entry.Values {
+				keys = append(keys, k)
+			}
+			sort.Strings(keys)
+
+			for _, k := range keys {
+				for _, v := range entry.Values[k] {
+					t := v.UnixNano()
+
+					// Check for duplicate/out of order keys.
+					if min, ok := minTimestampByKey[k]; ok && t <= min {
+						minTimestampByKey[k] = t
```

Comment:
`minTimestampByKey[k]` is not set unless the condition is true.
The condition is not true unless `minTimestampByKey[k]` has been set.

I haven't actually run the code, but I think this block is never hit.